### PR TITLE
Autocomplete: Surface rate limit and other errors

### LINKED ISSUE #634: Autocomplete: Expose rate limiting and other errors to users
There are some rate limits in place, both for chats and autocomplete. For chat, we display error messages in the transcript, so users know that their chat message failed because of rate limit exhaustion. But for autocomplete, we give no visible feedback.

We need to:
- Make it visible there's a rate limit happening
- Give them enough information to understand what's happening
- Know what they can do about it (wait, read the docs?)

From @eseliger’s [Slack message](https://sourcegraph.slack.com/archives/C052G9Y5Y8H/p1691020043358199):
> For autocompletions, as far as I know we don’t indicate errors too heavily today.
> We’ve been thinking of making the Cody navbar item turn red or something like that and on hover show it’s not available at the moment due to rate limits.

- [x] Design
- [x] Reviewed & ready to go

### PR DESCRIPTION
**Pull request by philipp-spiess**
Closes #634

This PR adds a UI indicator for autocomplete errors. We special case rate limiting errors to provider more information.

- When a rate limit error is encountered, we show it only once per rate limit interval (the UI indicator will reset with a VS Code restart). When the option is selected in quick pick, we open https://about.sourcegraph.com/blog/increasing-the-completions-rate-limit (the only resource I found that we have for rate limits)
- When any other (e.g. a network error) is encountered, we show the error message and open the output channel. We also add a new `error` API similar to the `debug` one but that will always log errors to the output channel. 

## ToDo

- [x] Add a new `error` API similar to the `debug` one but that will always log errors to the output channel. 
- [x] Add test cases

## Test plan

https://github.com/sourcegraph/cody/assets/458591/efa5f660-706e-47db-96e1-63b5f7b2d2fc

### CODE REVIEW

**Comment by philipp-spiess** [comment]
@toolmantim Please take a look at the attached video if this workflow makes sense for you. I've decided to add it to the same quick pick that the status bar item always opens (which is different from your UI sketches). The rest should be fairly similar.

File: vscode/src/services/StatusBar.ts [hunk 1]
```diff
@@ -6,16 +6,23 @@ import { getConfiguration } from '../configuration'
...
     startLoading(label: string): () => void
+    addError(error: StatusBarError): void
 }
 
 const DEFAULT_TEXT = '$(cody-logo-heavy)'
 const DEFAULT_TOOLTIP = 'Cody Settings'
 
 const QUICK_PICK_ITEM_CHECKED_PREFIX = '$(check) '
-const QUICK_PICK_ITEM_EMPTY_INDENT_PREFIX = '\u00A0\u00A0\u00A0\u00A0 '
+const QUICK_PICK_ITEM_EMPTY_INDENT_PREFIX = '\u00A0\u00A0\u00A0\u00A0\u00A0 '
```
Comment (philipp-spiess) [inline vscode/src/services/StatusBar.ts:25]: @toolmantim For me the existing indentation looked a bit too narrow:

### Before

Screenshot 2023-08-29 at 14 56 56

### After

Screenshot 2023-08-29 at 14 57 25

But maybe that's different for other themes?

File: vscode/src/completions/logger.ts [hunk 2]
```diff
@@ -1,13 +1,13 @@
 import { LRUCache } from 'lru-cache'
 import * as vscode from 'vscode'
 
+import { isAbortError, isRateLimitError } from '@sourcegraph/cody-shared/src/sourcegraph-api/errors'
```
Comment (valerybugakov) [inline vscode/src/completions/logger.ts:4]: 💜

File: vscode/src/completions/vscodeInlineCompletionItemProvider.ts [hunk 3]
```diff
@@ -161,86 +165,96 @@ export class InlineCompletionItemProvider implements vscode.InlineCompletionItem
...
+            }
+
+            const completionResult: vscode.InlineCompletionList = { items }
 
-        ;(completionResult as any).completionEvent = event
+            ;(completionResult as any).completionEvent = event
 
-        return completionResult
+            return completionResult
+        } catch (error: unknown) {
```
Comment (valerybugakov) [inline vscode/src/completions/vscodeInlineCompletionItemProvider.ts:254]: We can omit the `unknown` type here because we have [the `strict ` mode](https://www.typescriptlang.org/tsconfig#useUnknownInCatchVariables) enabled:

```suggestion
        } catch (error) {
```

File: vscode/src/completions/vscodeInlineCompletionItemProvider.ts [hunk 4]
```diff
@@ -295,6 +309,44 @@ export class InlineCompletionItemProvider implements vscode.InlineCompletionItem
...
+                    )
+                },
+            })
+            return
+        }
+
+        if (this.reportedErrorMessages.has(error.message)) {
+            return
+        }
+        this.reportedErrorMessages.add(error.message)
```
Comment (valerybugakov) [inline vscode/src/completions/vscodeInlineCompletionItemProvider.ts:341]: Does this mean that non-rate-limit errors will be shown only once per the activated extension lifetime? 

It would be helpful to show all the errors (even if a user already saw errors with this particular message) if they happen after a certain period. This way, we ensure we avoid flooding them with the same error messages and keep users informed about new errors occurring without swallowing them.

File: vscode/src/chat/MessageProvider.ts [hunk 5]
```diff
@@ -710,8 +710,8 @@ export abstract class MessageProvider extends MessageHandler implements vscode.D
                     void vscode.commands.executeCommand('vscode.open', exportPath)
                 }
             })
-        } catch (error) {
-            debug('MessageProvider:exportHistory', 'Failed to export chat history', error)
+        } catch (error_) {
+            error('MessageProvider:exportHistory', 'Failed to export chat history', error_)
```
[the remark this answers is not in the document]
Comment (philipp-spiess) [reply vscode/src/chat/MessageProvider.ts:714]: I’m fine with renaming but I'll have to also rename `debug` to `logDebug` because otherwise I can no longer sleep in peace 😅

File: vscode/src/services/StatusBar.ts [hunk 6]
```diff
@@ -143,12 +185,13 @@ export function createStatusBar(): CodyStatusBar {
...
                 openLoadingLeases--
-                if (openLoadingLeases === 0) {
-                    statusBarItem.text = DEFAULT_TEXT
-                    statusBarItem.tooltip = DEFAULT_TOOLTIP
-                }
+                rerender()
             }
         },
+        addError(error: StatusBarError) {
+            errors.push(error)
```
[the remark this answers is not in the document]
Comment (philipp-spiess) [reply vscode/src/services/StatusBar.ts:192]: Errors are removed when you _select_ them from the quick pick. Or do you mean cleaning them after a few minutes or so in case the user does not interact with it? hmm

File: vscode/src/chat/MessageProvider.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (valerybugakov) [reply vscode/src/chat/MessageProvider.ts:714]: 😃 thank you!

File: vscode/src/services/StatusBar.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (valerybugakov) [reply vscode/src/services/StatusBar.ts:192]: Here, I was referring to somehow cleaning the array with error objects. Can it grow indefinitely if I use Cody non-stop and do not restart my VS Code instance? 

Also, from the UX perspective, I love the idea of removing errors after a certain period. It's not helpful to see a message about the failed request from 30 mins ago 🤔

File: vscode/src/services/StatusBar.ts (hunk 1 above)
Comment (philipp-spiess) [reply vscode/src/services/StatusBar.ts:25]: Screenshot 2023-08-30 at 10 49 29
Should have added these lines right away so it's clearer what I mean :D

File: vscode/src/services/StatusBar.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (philipp-spiess) [reply vscode/src/services/StatusBar.ts:192]: We remove the error once you look at it, if you never look at them, yeah they can grow unbound theoretically. Do you think this is a problem?

What about we remove all errors after 1hr?

[the remark this answers is not in the document]
Comment (valerybugakov) [reply vscode/src/services/StatusBar.ts:192]: > What about we remove all errors after 1hr?

Yeah, a super simple solution like this works. I'm worried about memory leaks because debugging them as customer issues is tricky.

[the remark this answers is not in the document]
Comment (philipp-spiess) [reply vscode/src/services/StatusBar.ts:192]: I’m not worried about this, for that to happen they'd also have a completely flooded output tab (we now log errors even if debug logging is disabled) so it should be easy to find out what's going on.

[the remark this answers is not in the document]
Comment (valerybugakov) [reply vscode/src/services/StatusBar.ts]: > they'd also have a completely flooded output tab

Could it happen if a user has occasional errors and uses the extension for a few weeks without the VS Code instance?

File: vscode/src/completions/vscodeInlineCompletionItemProvider.ts [hunk 7]
```diff
@@ -295,6 +311,48 @@ export class InlineCompletionItemProvider implements vscode.InlineCompletionItem
...
+            }
+            this.resetRateLimitErrorsAfter = error.retryAfter?.getTime() ?? Date.now() + 24 * 60 * 60 * 1000
+            this.config.statusBar.addError({
+                title: 'Cody Autocomplete Disabled Due to Rate Limit',
+                description:
+                    `You've used all${error.limit ? ` ${error.limit}` : ''} daily autocompletions.` +
+                    (error.retryAfter ? ` Usage will reset in ${formatDistance(error.retryAfter, new Date())}.` : ''),
+                onSelect: () => {
+                    void vscode.env.openExternal(
+                        vscode.Uri.parse('https://about.sourcegraph.com/blog/increasing-the-completions-rate-limit')
```
Comment (toolmantim) [inline vscode/src/completions/vscodeInlineCompletionItemProvider.ts:333]:
```suggestion
                        vscode.Uri.parse('https://docs.sourcegraph.com/cody/troubleshooting#autocomplete-rate-limits')
```

**Review by toolmantim — APPROVED** [review]
Looks great!

I put up a quick docs addition in https://github.com/sourcegraph/sourcegraph/pull/56304 so we can link to there in stead of the blog.

File: vscode/src/services/StatusBar.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (philipp-spiess) [reply vscode/src/services/StatusBar.ts]: @valerybugakov Can you rephrase this? I’m confused what you mean by "uses the extension [...] without the VS Code instance"

**Comment by philipp-spiess** [comment]
@toolmantim @valerybugakov Quick update: I decided to only surface rate limiting errors for now until we resolve https://sourcegraph.slack.com/archives/C05AGQYD528/p1693473094573439?thread_ts=1693471486.690459&cid=C05AGQYD528 

This would be too much spam otherwise and definitely not a loveable experience
